plugins/wasi_nn/ggml.cpp

```diff
@@ -366,100 +356,134 @@ Expect<ErrNo> compute(WasiNNEnvironment &Env, uint32_t ContextId) noexcept {
-    spdlog::info("[WASI-NN] GGML backend: llama_decode() failed"sv);
-    return ErrNo::RuntimeError;
-  }
-
-  int NCur = LlamaBatch.n_tokens;
-  while (NCur < MaxContextSize && NCur < NPredict) {
-    // Sample the next token
-    auto NVocab = llama_n_vocab(GraphRef.LlamaModel);
-    auto *Logits =
-        llama_get_logits_ith(CxtRef.LlamaContext, LlamaBatch.n_tokens - 1);
-
-    std::vector<llama_token_data> Candidates;
-    Candidates.reserve(NVocab);
-    for (llama_token TokenId = 0; TokenId < NVocab; TokenId++) {
-      Candidates.emplace_back(llama_token_data{TokenId, Logits[TokenId], 0.0f});
-    }
-    llama_token_data_array CandidatesP = {Candidates.data(), Candidates.size(),
-                                          false};
-
-    // Sample the most likely token
-    const llama_token NewTokenId =
-        llama_sample_token_greedy(CxtRef.LlamaContext, &CandidatesP);
-
-    // Is it an end of stream?
-    if (NewTokenId == llama_token_eos(CxtRef.LlamaContext) ||
-        NCur == MaxContextSize || NCur == NPredict) {
-      break;
-    }
+  gpt_params GPTParams;
+  struct llama_sampling_context *CtxSampling =
+      llama_sampling_init(GPTParams.sparams);
+  std::vector<llama_token> Embd;
+  int NPast = 0;
+  int NConsumed = 0;
+  int NRemain = GraphRef.NPredict;
+  int NKeep = GPTParams.n_keep;
+  // Initialize the llama context.
+  llama_context_params ContextParams = llama_context_default_params();
+  ContextParams.n_ctx = GraphRef.CtxSize;
+  ContextParams.n_batch = GraphRef.BatchSize;
+  CxtRef.LlamaContext =
+      llama_new_context_with_model(GraphRef.LlamaModel, ContextParams);
+  int NCtx = llama_n_ctx(CxtRef.LlamaContext);
+  // Minus 4 for the special tokens.
+  const int MaxTokensListSize = NCtx - 4;
+  while (NRemain != 0) {
+    // Preidct
+    if (!Embd.empty()) {
+      // Truncate if necessary.
+      if (static_cast<int>(Embd.size()) > MaxTokensListSize) {
+        auto NSkipped = Embd.size() - MaxTokensListSize;
+        Embd.resize(MaxTokensListSize);
+        if (GraphRef.EnableLog) {
+          spdlog::info("[WASI-NN] GGML backend: Truncated {} tokens"sv,
+                       NSkipped);
+        }
+      }
 
-    std::string NextToken =
-        llama_token_to_piece(CxtRef.LlamaContext, NewTokenId);
+      // Infinite text generation via context swapping.
+      if (NPast + static_cast<int>(Embd.size()) > NCtx) {
+        const int NLeft = NPast + NKeep - 1;
+        const int NDiscard = NLeft / 2;
+        if (GraphRef.EnableLog) {
+          spdlog::info(
+              "[WASI-NN] GGML backend: Context full, swapping: NPast = {}, NLeft = {}, NKeep = {}, NDiscard = {}"sv,
+              NPast, NLeft, NKeep, NDiscard);
+        }
+        llama_kv_cache_seq_rm(CxtRef.LlamaContext, 0, NKeep + 1,
+                              NKeep + NDiscard + 1);
+        llama_kv_cache_seq_shift(CxtRef.LlamaContext, 0, NKeep + 1 + NDiscard,
+                                 NPast, -NDiscard);
```

Comment:
Signature: `(context, sequence id(always be 0), begin position, end position, delta)`.
This will shift `[NKeep+1+NDiscard, NPast)` for the `delta` distance.

Reply:
Ditto.